modules/sdk-core/src/account-lib/mpc/tss/ecdsa/ecdsa.ts

```diff
@@ -145,6 +147,102 @@ export default class Ecdsa {
+    const secretKey = seedWithValidLength && bip32.fromSeed(seedWithValidLength);
+    const u =
+      (secretKey && secretKey.privateKey && bigIntFromU8ABE(new Uint8Array(secretKey.privateKey))) ??
+      Ecdsa.curve.scalarRandom();
+    const y = Ecdsa.curve.basePointMult(u);
+
+    const chaincode = (secretKey && secretKey.chainCode) ?? randomBytes(32);
+
+    return {
+      i: index,
+      t: threshold,
+      c: numShares,
+      uu: u.toString(),
+      comDecomY: HashCommitment.createCommitment(bigIntToBufferBE(y, 33)),
+      chaincode: chaincode.toString('hex'),
+    };
+  }
+
+  /**
+   * Generate shares for participant at index and split keys `(threshold,numShares)` ways.
+   * @param {UYShare} uyShare share generated by keyShareStep1()
+   * @param sync optional sync flag, if true then a synchronous version of Paillier key generation is used that does not spawn Worker threads.
+   * @returns {Promise<KeyShare>} Returns the private p-share
+   * and n-shares to be distributed to participants at their corresponding index including decommitment to public y value.
+   */
+  async keyShareStep2(uyShare: UYShare, sync = false): Promise<KeyShare> {
+    // Generate additively homomorphic encryption key.
+    let paillierKeyPair: paillierBigint.KeyPair;
+    if (!sync) {
+      paillierKeyPair = await paillierBigint.generateRandomKeys(minModulusBitLength, true);
+    } else {
+      paillierKeyPair = paillierBigint.generateRandomKeysSync(minModulusBitLength, true);
+    }
+    const { publicKey, privateKey } = paillierKeyPair;
+
+    // Compute secret shares of the private key
+    const { shares: uShares, v } = Ecdsa.shamir.split(BigInt(uyShare.uu), uyShare.t, uyShare.c);
+    const currentParticipant: PShare = {
+      i: uyShare.i,
+      t: uyShare.t,
+      c: uyShare.c,
+      l: bigIntToBufferBE(privateKey.lambda, 192).toString('hex'),
+      m: bigIntToBufferBE(privateKey.mu, 192).toString('hex'),
+      n: bigIntToBufferBE(publicKey.n, 384).toString('hex'),
+      y: uyShare.comDecomY.decommitment.secret.toString('hex'),
+      u: bigIntToBufferBE(uShares[uyShare.i], 32).toString('hex'),
+      uu: uyShare.uu,
+      chaincode: uyShare.chaincode,
+    };
+    const keyShare: KeyShare = {
+      pShare: currentParticipant,
+      nShares: {},
+    };
+
+    for (const share in uShares) {
+      const participantIndex = parseInt(share, 10);
+      if (participantIndex !== uyShare.i) {
+        keyShare.nShares[participantIndex] = {
+          i: participantIndex,
+          j: currentParticipant['i'],
+          n: publicKey.n.toString(16),
+          y: uyShare.comDecomY.decommitment.secret.toString('hex'),
+          decomY: uyShare.comDecomY.decommitment,
+          v: bigIntToBufferBE(v[0], 33).toString('hex'),
+          u: bigIntToBufferBE(uShares[participantIndex], 32).toString('hex'),
+          chaincode: uyShare.chaincode,
+        } as NShare;
```

Comment:
Why do we need to cast here?

Reply:
It was copied over from the original keyShare(), both removed.